node/libs/storage/src/buffered.rs

```diff
@@ -0,0 +1,276 @@
+//! Buffered [`BlockStore`] implementation.
+
+use crate::{
+    block_store::{BlockStore, MissingBlockNumbers, WriteBlockStore},
+    StorageError, StorageResult,
+};
+use async_trait::async_trait;
+#[cfg(test)]
+use concurrency::ctx::channel;
+use concurrency::{
+    ctx,
+    sync::{self, watch, Mutex},
+};
+use roles::validator::{BlockNumber, FinalBlock};
+use std::{collections::BTreeMap, ops};
+
+/// [`BlockStore`] variation that upholds additional invariants as to how blocks are processed.
+///
+/// The invariants are as follows:
+///
+/// - Stored blocks always have contiguous numbers; there are no gaps.
+/// - Blocks can be scheduled to be added using [`Self::schedule_next_block()`] only. New blocks do not
+///   appear in the store otherwise.
+#[async_trait]
```

Comment:
Are you sure this is the right place for specifying this abstraction? AFAIU it is supposed to be the wrapper for postgres in zksync-era, and the consensus repo doesn't really have any direct use for it. How about moving ContiguousBlockStore and BufferedStorage to zksync-era? And perhaps we won't need to make it generic then?

Reply:
I thought about `BufferedStorage` as a helper that is generic enough that it can be placed close to functionality that provides (i.e., `BlockStore` / `WriteBlockStore` traits), as a helper. In general terms, it just expresses sequential block execution, which is not specific to `zksync-era` / Postgres-based DAL. Does this make sense, or not quite? Regarding generics, I wouldn't remove them in any case because they make `BufferedStorage` easier to test.